### PR TITLE
Fix FY5253 onOffset/apply bug, simplify

### PR DESCRIPTION
**Pull request by jbrockmendel**
Similar to #18875.

The actual bugfix here is just changing the comparison in `apply` from `other < whatever` to `norm < whatever`.  The remaining edits to `get_year_end` are orthogonal simplification (which I guess could go in a separate PR).

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tseries/offsets.py [hunk 1]
```diff
@@ -1821,33 +1816,27 @@ def apply(self, other):
...
+            if norm < prev_year:
                 n -= 2
-            elif prev_year < other < cur_year:
+            elif prev_year < norm < cur_year:
                 n -= 1
-            elif cur_year < other < next_year:
+            elif cur_year < norm < next_year:
                 pass
-            else:
-                assert False
```
Comment (jbrockmendel) [inline pandas/tseries/offsets.py]: Double-checked this, it cannot be reached.   See new comment above re `next_year.year`.

File: pandas/tseries/offsets.py [hunk 2]
```diff
@@ -1821,33 +1816,27 @@ def apply(self, other):
...
         else:
-            if next_year < other:
-                n += 2
-                # TODO: Not hit in tests; UPDATE: looks impossible
```
Comment (jbrockmendel) [inline pandas/tseries/offsets.py]: Double-checked this, it cannot be reached.   See new comment above re `next_year.year`.

File: pandas/tseries/offsets.py [hunk 3]
```diff
@@ -1863,35 +1852,29 @@ def apply(self, other):
...
-            return self._get_year_end_nearest(dt)
-        else:
-            return self._get_year_end_last(dt)
+        assert dt.tzinfo is None
 
-    def get_target_month_end(self, dt):
-        target_month = datetime(dt.year, self.startingMonth, 1,
-                                tzinfo=dt.tzinfo)
-        return shift_month(target_month, 0, 'end')
-        # TODO: is this DST-safe?
```
Comment (jbrockmendel) [inline pandas/tseries/offsets.py]: non-issue b/c/ get_target_month_end is only ever called with `dt.tzinfo is None`

**Comment by jbrockmendel** [comment]
That should be the last new one for a while...  I hope.

**Comment by jbrockmendel** [comment]
Close was accidental.

File: pandas/tests/tseries/offsets/test_fiscal.py [hunk 4]
```diff
@@ -158,17 +158,6 @@ def test_apply(self):
 
 class TestFY5253NearestEndMonth(Base):
 
-    def test_get_target_month_end(self):
-        assert (makeFY5253NearestEndMonth(
```
Comment (jreback) [inline pandas/tests/tseries/offsets/test_fiscal.py]: are these tests wrong?

File: pandas/tests/tseries/offsets/test_fiscal.py [hunk 5]
```diff
@@ -625,3 +614,20 @@ def test_bunched_yearends():
     assert fy.rollback(dt) == Timestamp('2002-12-28')
     assert (-fy).apply(dt) == Timestamp('2002-12-28')
     assert dt - fy == Timestamp('2002-12-28')
+
+
+def test_fy5253_last_onoffset():
+    offset = FY5253(n=-5, startingMonth=5, variation="last", weekday=0)
```
Comment (jreback) [inline pandas/tests/tseries/offsets/test_fiscal.py]: add a comment to these tests about what you are testing. next person to come along will have no idea. add the issue number as well.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.22.0.txt [hunk 6]
```diff
@@ -354,3 +354,4 @@ Other
 
 - Improved error message when attempting to use a Python keyword as an identifier in a ``numexpr`` backed query (:issue:`18221`)
 - Bug in :class:`Timestamp` where comparison with an array of ``Timestamp`` objects would result in a ``RecursionError`` (:issue:`15183`)
+- Bug in :class:`FY5253` where ``datetime`` addition and subtraction incremented incorrectly in some corner cases (:issue:`18854`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: can you give a typical case here, so a user would know to look closer about this.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/tseries/offsets/test_fiscal.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/tseries/offsets/test_fiscal.py]: No, but `get_target_month_end` is removed in this PR.

File: pandas/tseries/offsets.py [hunk 7]
```diff
@@ -1821,33 +1816,27 @@ def apply(self, other):
         prev_year = tslib._localize_pydatetime(prev_year, other.tzinfo)
         cur_year = tslib._localize_pydatetime(cur_year, other.tzinfo)
         next_year = tslib._localize_pydatetime(next_year, other.tzinfo)
+        # Note: next_year.year == other.year + 1, so we will always
```
Comment (jreback) [inline pandas/tseries/offsets.py]: pls add a blank line (generally) before comments.

File: pandas/tseries/offsets.py [hunk 8]
```diff
@@ -1863,35 +1852,29 @@ def apply(self, other):
...
+            # year_end is the same for "last" and "nearest" cases
             return target_date
 
-        days_forward = wkday_diff % 7
-        if days_forward <= 3:
-            # The upcoming self.weekday is closer than the previous one
-            return target_date + timedelta(days_forward)
+        if self.variation == "last":
+            days_forward = (wkday_diff % 7) - 7
+            # days_forward is always negative, so we always end up
```
Comment (jreback) [inline pandas/tseries/offsets.py]: here

File: pandas/tseries/offsets.py [hunk 9]
```diff
@@ -1863,35 +1852,29 @@ def apply(self, other):
...
-            return target_date + timedelta(days_forward - 7)
-
-    def _get_year_end_last(self, dt):
-        current_year = datetime(dt.year, self.startingMonth, 1,
-                                tzinfo=dt.tzinfo)
-        return current_year + self._offset_lwom
+            # variation == "nearest":
+            days_forward = wkday_diff % 7
+            if days_forward <= 3:
+                # The upcoming self.weekday is closer than the previous one
```
Comment (jreback) [inline pandas/tseries/offsets.py]: these comments are good, FYI (you only need a blank line when the indentation is the same)

**Review by jreback — CHANGES_REQUESTED** [review]
minor formatting. ping on green.

**Comment by jbrockmendel** [comment]
Ping

**Comment by jbrockmendel** [comment]
This is a blocker for addressing bugs in FY5253Quarter.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. conflict in whatsnew. ping when pushed / green.

**Comment by jbrockmendel** [comment]
ping

**Comment by jreback** [comment]
thanks!

**Pull request merged**